src/impression.js

```diff
@@ -164,6 +164,7 @@ function handleClickUrl(win) {
     return Promise.resolve();
   }
 
+  ////////////////// TODO: set ampshare without this
```

Comment:
Also note that `replaceState` is called below in `applyResponse()`.

I don't really get what the point of this class is. AMP ALP? Why change the window location from the result of a JSON fetch to the "click" viewer param?  @zhouyx for context.

Reply:
change window location from the result of a JSON fetch is for tracking purpose. We change the window location so that `<amp-analytics>` and `<amp-pixel>` can query params. 

`click` viewer param contains the url for tracking. Instead of navigating to `track.com` to track conversion and then to AMP landing page. A better way is to navigate to landing page directly with `track.com` as the `click` param. Then ask landing page to track conversion by pinging `track.com`

Reply:
Thanks!

Reply:
I think I've covered all cases, but a careful review would be much appreciated.

 - For Replace URLs, I think the logic should be fine as is since it appears to preserve the fragment.
 - For Click URLs, I make sure that the ampshare fragment is restored after pinging the click URL
   - There are several branches in click URL code that would cause the ampshare param to be lost, but since these are unlikely to occur -- and the failure mode isn't a big deal -- it seems fine to ignore these

One other thing to consider: setting the ampshare fragment will clear any existing fragment right now. However, it seems the only use for fragments before this point is mode.js (e.g., to set "development" and "filter" flags).

Since these fragments are currently captured before we set the fragment (they're copied to win.location.originalHash), it seems safe to repurpose the fragment.

What do you think?

Reply:
I'm not sure that click URLs won't have hashes. You'd overwrite it if they do.